### ISSUE TITLE
Add a DirectoryPath property to DirectoryNotFoundException to be consistent with FileNotFoundException

### ISSUE THREAD

**Issue by daiplusplus** [dotnet/runtime#103468]
### Background and motivation

[The `DirectoryNotFoundException` type](https://learn.microsoft.com/en-us/dotnet/api/system.io.directorynotfoundexception?view=net-8.0) does not provide any way for the thrower to specify the actual directory path in a machine-readable way such that any downstream/consuming program code that catches the exception can isolate the directory-path; [currently we're forced to try to parse the human-readable `.Message` property](https://stackoverflow.com/questions/6460512/c-sharp-get-the-directory-name-from-the-directorynotfoundexception) which, additionally, might be a localized string and thus impossible to reliably process under all circumstances, which is important in error-handling functions.

It's also inconsistent with how `FileNotFoundException` has [a `String FileName { get; }` property](https://learn.microsoft.com/en-us/dotnet/api/system.io.filenotfoundexception.filename?view=net-8.0).

----

I did search [through all 7 pages of existing Issues in the runtime repo matching "`DirectoryNotFoundException`"](https://github.com/dotnet/runtime/issues?page=7&q=is%3Aissue+directorynotfoundexception) - none of them are a request to add a directory-path property to the exception.

### API Proposal

```csharp
namespace System.IO;

public class DirectoryNotFoundException : IOException
{
    public DirectoryNotFoundException (string? message, string? directoryPath, Exception? innerException)
        : this(message: message, innerException: innerException)
    {
        this.DirectoryPath = directoryPath;
    }

    public DirectoryNotFoundException (string? message, string? directoryPath)
        : this(message: message)
    {
        this.DirectoryPath = directoryPath;
    }

    public string? DirectoryPath { get; }
}
```

### API Usage

```csharp

public static void ExampleMethodThatCopiesFilesAround(DirectoryInfo source, DirectoryInfo dest)
{
    if (!source.Exists) throw new DirectoryNotFoundException(message: "Source directory does not exist.", directoryPath: source.FullName);
    if (!dest.Exists) throw new DirectoryNotFoundException(message: "Destination directory does not exist.", directoryPath: dest.FullName);

    foreach (FileInfo fi in source.GetFiles())
    {
        // etc...
    }
}

//

using Microsoft.Extensions.Logging;

public void ConsumingMethodWithLogger()
{
    try
    {
        ExampleMethodThatCopiesFilesAround(source: foo, dest: bar);
    }
    catch (DirectoryNotFoundException dEx)
    {
        using (this.logger.BeginScope("DirectoryPath", dEx.DirectoryPath))
        {
            this.logger.LogError(exception: dEx, message: "Directory not found: {DirectoryPath}", dEx.DirectoryPath);
        }
    }
}
```

### Risks

This change involves only adding additional (and entirely optional and nullable) constructor overloads and properties.

It's _possible_ that there exists user-code that subclasses `DirectoryNotFoundException` to add their own property for the missing directory's full-path, but I assume this wouldn't be an ABI breaking change because user subclasses would still be using a different class member "slot" (right?)

**Comment by huoyaoyuan** [comment]
There are situations that the directory path can't be determined at all. For example on Windows, trying to open a file from an non-existing directory. The OS error doesn't tell about which level of directory is missing.

**Comment by julealgon** [comment]
> There are situations that the directory path can't be determined at all. For example on Windows, trying to open a file from an non-existing directory. The OS error doesn't tell about which level of directory is missing.

@huoyaoyuan the same situation occurs with `HttpRequestException` where a status code is not always present. In that case, the `StatusCode` property was made nullable.

Same would apply here.

**Comment by terrajobst** [comment]
* We discussed making the property non-null but decided it was simpler to just store what we were given

```C#
namespace System.IO;

public partial class DirectoryNotFoundException
{
    // Existing ctors
    // public DirectoryNotFoundException();
    // public DirectoryNotFoundException(String);
    // public DirectoryNotFoundException(SerializationInfo, StreamingContext);
    // public DirectoryNotFoundException(String, Exception);
    public DirectoryNotFoundException(string? message, string? directoryPath);
    public DirectoryNotFoundException(string? message, string? directoryPath, Exception? innerException);
    public string? DirectoryPath { get; }
}
```

**Comment by daiplusplus** [comment]
> If `message` is null but a directoryPath is given Message should return a constructed one, similar to `FileNotFoundException`

I did consider that when I wrote-up my initial proposal, but I was wary of the risk of unintentional information disclosure (I understand this is the same reason why `KeyNotFoundException` doesn't expose the _key-that-wasn't-found_). But if you're fine with that then I'm fine with that.

**Comment by daiplusplus** [comment]
> @terrajobst If message is null but a `directoryPath` is given `Message` should return a constructed one, similar to `FileNotFoundException`

I've started my implementation and hit a minor problem with the `.Message` generation part: I was modelling my impl off `FileNotFoundException` - but its `.Message` string factory ends-up being a call into the CLR via `FileLoadException.FormatFileLoadExceptionMessage`.

Should I just write my own equivalent message strings in `SR` then?

**Comment by terrajobst** [comment]
@jozkee should be able to guide you

**Comment by jozkee** [comment]
> Should I just write my own equivalent message strings in SR then?

Yes. Feel free to send a draft PR and we can further discuss the approach.

**Comment by Flu** [comment]
Is anybody working on this currently? It seems not and nobody has commented or made any PRs addressing this issue.

I can start working on it if that's okay with you, @jozkee

**Comment by ViveliDuCh** [comment]
Hi! Just checking in since it’s been over a year since the last update. @daiplusplus, are you still planning to submit a draft PR? If not, I would like to pick this up. Of course, happy to step aside if you’re still working on it. Thanks!
